src/Network/GRPC/MQTT/Client.hs

```diff
@@ -209,78 +194,112 @@ mqttRequest MQTTGRPCClient{..} baseTopic nmMethod useBatchedStream request = do
-        timeoutGRPC timeLimit $ do
-          publishRequest Proto3.def timeLimit initMetadata
-          withControlSignals publishControlMsg . exceptToResult $ do
-            liftIO $ do
-              -- do client streaming
-              streamHandler publishToRequestStream
-              -- Send end of stream indicator
-              publishToStreamCompleted
-
-            rsp <- withExceptT parseErrorToRCE (packetReader responseChan)
-            liftEither $ unwrapClientStreamResponse rsp
-
-      -- Server Streaming Requests
-      MQTTReaderRequest req timeLimit reqMetadata streamHandler ->
-        timeoutGRPC timeLimit $ do
-          publishRequest req timeLimit reqMetadata
-          withControlSignals publishControlMsg . exceptToResult $ do
-            -- Wait for initial metadata
-            rawInitMetadata <- withExceptT parseErrorToRCE (packetReader responseChan)
-            metadata <- liftEither $ parseMessageOrError rawInitMetadata
-
-            readResponseStream <- mkStreamRead $ withExceptT parseErrorToRCE (packetReader responseChan)
-
-            let mqttSRecv :: StreamRecv response
-                mqttSRecv = runExceptT . withExceptT toGRPCIOError $ readResponseStream
-
-            -- Run user-provided stream handler
-            liftIO (streamHandler (toMetadataMap metadata) mqttSRecv)
-
-            -- Return final result
-            rsp <- withExceptT parseErrorToRCE (packetReader responseChan)
-            liftEither $ unwrapServerStreamResponse rsp
-
-      -- BiDirectional Server Streaming Requests
-      MQTTBiDiRequest timeLimit reqMetadata streamHandler ->
-        timeoutGRPC timeLimit $ do
-          publishRequest Proto3.def timeLimit reqMetadata
-          withControlSignals publishControlMsg . exceptToResult $ do
-            -- Wait for initial metadata
-            rawInitMetadata <- withExceptT parseErrorToRCE (packetReader responseChan)
-            metadata <- liftEither $ parseMessageOrError rawInitMetadata
-
-            readResponseStream <- mkStreamRead $ withExceptT parseErrorToRCE (packetReader responseChan)
-            let mqttSRecv :: StreamRecv response
-                mqttSRecv = runExceptT . withExceptT toGRPCIOError $ readResponseStream
-
-            let mqttSSend :: StreamSend request
-                mqttSSend = publishToRequestStream
-
-            let mqttWritesDone :: WritesDone
-                mqttWritesDone = do
-                  publishToStreamCompleted
-                  return $ Right ()
-
-            -- Run user-provided stream handler
-            liftIO $ streamHandler (toMetadataMap metadata) mqttSRecv mqttSSend mqttWritesDone
-
-            -- Return final result
-            rsp <- withExceptT parseErrorToRCE (packetReader responseChan)
-            liftEither $ unwrapBiDiStreamResponse rsp
+    let timeout :: TimeoutSeconds
+        timeout = requestTimeout request
+     in timeoutGRPC timeout case request of
```

Comment:
You can save a nesting level here by dropping the `in` since you are already inside a `do` block.